lib/Dialect/TritonGPU/Transforms/OptimizeGatherLayouts.cpp

```diff
@@ -0,0 +1,159 @@
+  SmallVector<unsigned> warpsPerCTA(rank);
+  SmallVector<unsigned> order;
+  order.push_back(axis);
+
+  // Minimize `sizePerThread[axis]` by putting as many theads along the axis as
+  // possible, limited to the actual size of the dimension.
+  unsigned maxThreadsInAxis =
+      std::min<unsigned>(srcType.getDimSize(axis), numThreadsPerWarp);
+  threadsPerWarp[axis] = maxThreadsInAxis;
+
+  // Now spread them along the other dimensions. Do this according to order
+  // (arbitrary).
+  unsigned threadsToAlloc = numThreadsPerWarp / maxThreadsInAxis;
+  auto distributedItf = cast<DistributedEncodingTrait>(srcType.getEncoding());
+  for (unsigned dim : distributedItf.getThreadOrder()) {
+    if (dim == axis)
+      continue;
+    // The gather axis is now the fastest-changing dimension.
+    order.push_back(dim);
+    unsigned nextThreadAlloc =
+        std::min<unsigned>(srcType.getDimSize(dim), threadsToAlloc);
+    threadsPerWarp[dim] = nextThreadAlloc;
+    threadsToAlloc /= nextThreadAlloc;
+  }
+  assert(llvm::none_of(threadsPerWarp, [](unsigned c) { return c == 0; }));
+
+  // There must be one warp along the gather axis.
+  warpsPerCTA[axis] = 1;
+  // Allocate the remaining warps in the same manner.
+  unsigned warpsToAlloc = numWarps;
+  for (unsigned dim : distributedItf.getWarpOrder()) {
+    if (dim == axis)
+      continue;
+    unsigned warpsCanFit = srcType.getDimSize(dim) / threadsPerWarp[dim];
+    assert(warpsCanFit != 0);
+    unsigned nextWarpAlloc = std::min<unsigned>(warpsCanFit, warpsToAlloc);
+    warpsPerCTA[dim] = nextWarpAlloc;
+    warpsToAlloc /= nextWarpAlloc;
+  }
+  assert(llvm::none_of(warpsPerCTA, [](unsigned c) { return c == 0; }));
+
+  // Just set `sizePerThread` to 1 along other dimensions and let broadcasting
+  // handling it. This also means we can use the same layout between the source
+  // and index tensors for simplicity.
+  SmallVector<unsigned> sizePerThread(rank, 1);
+  sizePerThread[axis] = srcType.getDimSize(axis) / threadsPerWarp[axis];
+
+  // Overflow by broadcasting along the gather axis since this is the most
+  // predictable.
+  threadsPerWarp[axis] *= threadsToAlloc;
+  warpsPerCTA[axis] *= warpsToAlloc;
+
+  assert(product(threadsPerWarp) == numThreadsPerWarp);
+  assert(product(warpsPerCTA) == numWarps);
+
+  // Construct the new layout.
+  MLIRContext *ctx = srcType.getContext();
+  auto ctaLayout = CTALayoutAttr::get(ctx, distributedItf.getCTAsPerCGA(),
+                                      distributedItf.getCTASplitNum(),
+                                      distributedItf.getCTAOrder());
+  auto newLayout = BlockedEncodingAttr::get(ctx, sizePerThread, threadsPerWarp,
+                                            warpsPerCTA, order, ctaLayout);
+
+  // Update the layout on the gather op and insert conversions.
+  mlir::ImplicitLocOpBuilder b(op.getLoc(), op);
+  auto cvtSrc = b.create<ConvertLayoutOp>(replaceEncoding(srcType, newLayout),
```

Comment:
Have you benchmarked and check its actually efficiency? I'm a bit worried about new the overhead of convert layouts being used here.

Reply:
The idea is that the layout conversions are removed/minimized by `tritongpu-remove-layout-conversions`. This isn't always possible, but with some of the work @lezcano is doing with improving layout propagation, the compiler actually does a pretty decent job at removing most of the conversions, and indeed this ends up being faster on the cases I was testing than using the shared memory version. It was good enough that I removed the heuristic in the function that prevent it from applying in certain cases, but we can certainly bring them back if necessary.

Reply:
I see. You added it before the rematerialization pass and assumed that that pass will be able to remove those convert layout ops

Reply:
Yes, and it usually does a pretty decent job. We can always check this in the backend and merge the layout conversions back into the gather if they didn't go away.